Add Basic Unit Tests

thiagomacg opened the pull request:
Boa tarde Alan, 

Não sei se assim seria a melhor maneira de fazer os comentários e correções : 

Bem, continuo fazendo os unit-test. 

Do Request anterior ainda ficou pendente : 
   > Revisar porque se a == b não funciona e b == a funciona. 
   > Passar o código pelo clang-format. Fiz a instalação, mas não consegui utilizar/configurar ainda. 

Neste Request estou com um erro que não to conseguindo resolver. 

No TEST ( logA.evaluate_sym(bool_val,int_val,doub_val) == logA ), onde logA = log(90,180)

Por fim, este teste apresenta um erro ao coletar a seguinte expressão :  1*1*4.49981. 

A expressão possui dois child_nodes : 1*1 e 4.49981

A função collect então deve separar o produto, inserir em child_nodes e depois apagar o termo que era um produto. 

O erro ocorre justamente ao tentar apagar o termo ( 1*1), na linha 278 ( child_nodes_.erase(i);) em product.cpp. 

Conferi em vários pontos e esta operação é usada frequentemente. Também tentei copiar o iterador e resolver de outras formas, mas nada deu certo. 

Poderia me ajudar neste ponto. 

Obrigado

thiagomacg committed:
Unit tests

thiagomacg committed:
Unit tests - Error Product Collect()

alandefreitas commented:
Olá Thiago,

Esta maneira de fazer as correções é bem melhor porque tenho acesso aos arquivos que foram alterados, aos resultados dos unit tests e posso fazer comentários direto no código sem ficar copiando e colando. 

Só não precisa criar PRs novos, a discussão toda pode ficar em um PR só. Também é importante colocar um nome no PR que diga o que ele resolve ou tenta resolver, em nome de erros. Erros devem ir para os issues. Em vez de mandar PRs grandes que não funcionam ou ficam em discussão por semanas, tente enviar PRs menores com tópicos bem definidos. Por exemplo, "Unit Test - product" ou "Fix product comparison" e enviar só os unit tests para um tipo de nó. 

Tente colocar os unit tests que já funcionam em um PR e vamos dar merge nesses para que a coisa possa andar. Depois vamos fazendo outros menores com tópicos mais bem definidos. Se o unit test passar, você pode mandar um PR "Unit Test - X", onde X é a coisa que está testando. Se o unit test não passar e precisar consertar algo, pode enviar um PR "Fix X", onde X é o nome da coisa que consertou e nos unit tests você já inclui um teste.

Sobre o clang-format, até onde sei, ele já vem no CLion. Dê uma pesquisada na documentação no google "clang-format clion". Pelo o que lembro, sempre que tem um arquivo ".clang-format.yml" na raíz do projeto, o próprio CLion começa a usar o clang format em vez das regras internas para formatar o código. Se clicar com o botão direito em uma pasta, ele dá a opção de reformatar todo o código ali:

<img width="298" alt="image" src="https://user-images.githubusercontent.com/5369819/96749740-92e3d800-13a1-11eb-8cd8-df288face844.png">

E quando vai dar o commit, ele também da a opção de já reformatar todo o código sendo enviado:

<img width="352" alt="image" src="https://user-images.githubusercontent.com/5369819/96750310-2fa67580-13a2-11eb-939e-bb40c11fbbfe.png">

No Linux ele me mostra as mesmas opções. Isso é importante não só porque garante que o código inteiro siga um formato e fique fácil de ler mas também porque evitar que os PRs fiquem cheios de arquivos supostamente alterados mas que na verdade são só espaços e quebras de linha que não mudam nada.

alandefreitas commented on source/sympp/core/sym.cpp:
```diff
@@ -680,7 +680,7 @@ namespace sympp {
     }
 
-    const std::type_info &sym::type() const { return this->root_node_->type(); }
+    const std::type_info &sym::type() const {return this->root_node_->type(); }
```
Esse tipo de coisa que o clang-format resolveria para nós.

thiagomacg replied:
Boa tarde, 

Qual a versão do CLion que esta utilizando ? 

Estou na 2.17.3  e Ubuntu 16.04 e não vem o clang-format dessa maneira. 
Tentei add seguindo como external-tools mas não funcionou corretamente https://stackoverflow.com/questions/34648255/using-clang-format-in-clion 

O Clion 2.17.3 dá opção de update para 2.17.5. Até atualizei mas desconfigurou o ambiente.

alandefreitas replied:
Sempre uso a mais recente. Não tem motivo para não atualizar o CLion sempre. 
Ele suporta clangformat há alguns anos já. As instruções do site são bem diretas:
https://www.jetbrains.com/help/clion/clangformat-as-alternative-formatter.html

Ubuntu 16.04 está bem antigo. Até as máquinas virtuais de teste usam ubuntu 18 se não me engano. 
Não sei se clang-format já tem que vir com Ubuntu ou se o CLion já tem ele embutido.
Na pior das hipóteses, está a um apt-get de distância.

thiagomacg replied:
Atualizei o CLion e já consigo utilizar clang-format.

alandefreitas commented on source/sympp/functions/mathematics.h:
```diff
@@ -22,9 +22,9 @@ namespace sympp {
     sym csc(const sym &);
 
-    sym sinh(const sym &);
+    sym Sinh(const sym &);
```
Coloque um alias a mais que só direciona para a outra função, mas deixe a versão antiga.

alandefreitas commented on source/sympp/functions/mathematics.h:
```diff
@@ -22,9 +22,9 @@ namespace sympp {
+    sym Sinh(const sym &);
 
-    sym cosh(const sym &);
+    sym Cosh(const sym &);
```
Mesmo

alandefreitas commented on source/sympp/node/function/abs.cpp:
```diff
@@ -26,24 +26,48 @@ namespace sympp {
     double abs::evaluate(const std::vector<uint8_t> &bool_values,
                          const std::vector<int> &int_values,
                          const std::vector<double> &double_values) const {
+
```
clang-format resolveria isso

alandefreitas commented on source/sympp/node/function/abs.cpp:
```diff
@@ -26,24 +26,48 @@ namespace sympp {
     sym abs::evaluate_sym(const std::vector<uint8_t> &bool_values,
                           const std::vector<int> &int_values,
                           const std::vector<double> &double_values) const {
+
```
e isso

thiagomacg replied:
Resolvido

alandefreitas commented on source/sympp/node/function/abs.cpp:
```diff
@@ -26,24 +26,48 @@ namespace sympp {
+
         auto ce = this->child_nodes_.front().root_node()->evaluate_sym(
-            bool_values, int_values, double_values);
+                bool_values, int_values, double_values);
```
e isso

alandefreitas commented on source/sympp/node/function/abs.cpp:
```diff
@@ -26,24 +26,48 @@ namespace sympp {
+                }
+
+            }
+
```
Não é necessário tratar os outros tipos de números? rational e boolean?

alandefreitas commented on source/sympp/node/function/abs.cpp:
```diff
@@ -26,24 +26,48 @@ namespace sympp {
+
+            if(ce.is_integer_number()){
+
+                if (static_cast<double>(p->operator int ()) < 0) {
```
Se p é inteiro, isso é ou deveria ser o mesmo que static_cast<double>(p)
`p < 0` já deveria funcionar aqui.

thiagomacg replied:
Não é possivel realizar a conversão static_cast(p), p é number_interface e gera falha no teste. 
Atualizado para apenas if (p->operator int() < 0)

alandefreitas replied:
O `p` é number_interface mas todos os number_interfaces tem os operadores:

```
       /// Convert the node to bool (might lose precision)
        virtual explicit operator bool() const;

        /// Convert the node to int (might lose precision)
        virtual explicit operator int() const;

        /// Convert the node to double (might lose precision)
        virtual explicit operator double() const = 0;
```

definidos, tanto que vc usa eles na próxima linha 47.

E mesmo isso seja problema, o `ce` não é number interface, bastaria usar ele direto, que também tem o operator operator int() definido na linha 418:

```cpp
/// Convert to bool
        explicit operator bool() const;

        /// Convert to int
        explicit operator int() const;

        /// Convert to double
        explicit operator double() const;
```

alandefreitas commented on source/sympp/node/function/abs.cpp:
```diff
@@ -26,24 +26,48 @@ namespace sympp {
+
             }
+
+            if(ce.is_real_number()){
```
Este código faz o mesmo que o de cima. Estas conversões, erros a parte, já estão definidas nos objetos.

alandefreitas commented on source/sympp/node/function/cos.cpp:
```diff
@@ -112,6 +112,8 @@ namespace sympp {
     std::optional<sym> cos::simplify(double, complexity_lambda) {
         sym &s = child_nodes_.front();
         s.simplify();
+
```
clang-format

alandefreitas commented on source/sympp/node/function/cosh.cpp:
```diff
@@ -116,6 +116,10 @@ namespace sympp {
             if (s.is_zero()) {
                 return sym(integer(1));
             }
+            if (s.is_integer_number()) {
```
Este trecho está fazendo o mesmo que o de baixo.

alandefreitas commented on source/sympp/node/function/log.cpp:
```diff
@@ -289,25 +292,25 @@ namespace sympp {
         }
 
         // log_c(b) -> log(c) * ln(b)^-1
-        if (a.is_number() && a.is_real_number() && a.operator double() > 0.0) {
```
Porque o operator double não funciona direto no sym?

alandefreitas commented on source/sympp/node/function/log.cpp:
```diff
@@ -289,25 +292,25 @@ namespace sympp {
-        if (a.compare(constant::e()) == 0 && b.is_number()) {
-            product p(sym(real(std::log(b.operator double()))),
+        if (constant::e().compare(a) == 0 && b.is_number()) {
+            product p(sym(real(std::log(b.root_node_as<number_interface>()->operator double()))),
```
Mesmo

thiagomacg replied:
O erro ocorria na comparação entre o simbolo e a constant e : 
A constante não converte dinamicamente para number interface. 
```

int number_interface::compare(const node_interface &node) const {

 if (node.type() == typeid(constant)) {
            const auto &cons = dynamic_cast<const constant &>(node);
            return compare_number(real(cons.value()));
        } else {
            const auto &p = dynamic_cast<const number_interface &>(node);
            return compare_number(p);
       }

} 
```


Retornadas as comparações para a forma inicial

alandefreitas commented on source/sympp/node/function/log.cpp:
```diff
@@ -323,7 +326,7 @@ namespace sympp {
     void log::stream(std::ostream &os, bool symbolic_format) const {
         if (child_nodes_.size() == 2 &&
-            child_nodes_.back().compare(constant::e()) == 0) {
+            constant::e().compare(child_nodes_.back()) == 0) {
```
O operador de igualdade deveria funcionar dos dois lados.

alandefreitas commented on source/sympp/node/function/pow.cpp:
```diff
@@ -385,7 +387,7 @@ namespace sympp {
             }
         } else {
-            if (child_nodes_.front().compare(constant::e()) == 0) {
+            if (constant::e().compare(child_nodes_.front()) == 0) {
```
Mesmo. Está evitando um erro em vez de consertar.

alandefreitas commented on source/sympp/node/function/sinh.cpp:
```diff
@@ -117,6 +117,10 @@ namespace sympp {
             if (s.is_zero()) {
                 return sym(integer(0));
             }
+            if (s.is_integer_number()) {
```
Faz exatamente que o trecho abaixo. Tem que pensar muito sempre se não tem algo errado quando tem dois trechos de código repetidos assim. 95% das vezes tem. As poucas vezes onde isto está correto é quando precisamos lidar com algum problema do compilador.

thiagomacg replied:
Aqui, havia apenas a comparação para número real, não sendo possível nos casos onde o número é inteiro executar std::sin(p->operator double()) 

Incluídos os tratamentos para booleano e racional. 

Retiradas as repetições e para Sin e Sinh retirado o caso base
 
```
if (s.is_zero()) {
         return sym(integer(0));
}, 

```
Este caso fica definido com os demais números inteiros.

alandefreitas replied:
Não entendi o que quer dizer com "havia apenas a comparação para número real", mas a class integer tem sim o operador "p->operator double()" definido. 

A diferenciação poderia fazer mais sentido se precisasse do "p->operator int()", mas não é o caso, pois na linha 122, vc usa o p->operator int() e logo depois converte o resultado para "real".

alandefreitas commented on source/sympp/node/operation/product.cpp:
```diff
@@ -263,8 +264,12 @@ namespace sympp {
         auto i = child_nodes_.begin();
         while (i != child_nodes_.end()) {
+
+            std::cout << std::endl << " *i " << *i << std::endl;
```
*Nunca nunca nunca* dê PR em código com anotações pessoais.

alandefreitas reviewed (changes requested):
Por favor dê PR em coisas menores para poder mover com o código. É muito maçante ter que revisar um código longo onde todos os erros são exatamente do mesmo tipo e a maioria deles já tinha comentado na última revisão. Tem que decidir se vai consertar ou não. Se não vai, melhor dar rollback no arquivo do que deixar o erro em todos os PRs.

alandefreitas commented:
Tem que reduzir esse PR. Várias coisas aí já poderia ter dado merge. Mas tem um monte de erro no meio que impede o merge.

As coisas que funcionam vc já pode enviar em um PR para merge e acabamos com isso. As coisas que não funcionam vc pode até abrir um draft mas não podem acumular. Se encontra um erro, tem que resolver ou não resolver. Não faz sentido ficar acumulando mais erro e enviar um PR com 14 arquivos alterados colocando mais erro em cima do erro que não foi corrigido.

Fica impossível revisar o código assim. Tem erro neste PR que é exatamente o erro do PR antigo. E tem erro aí que não tem relação nenhuma um com o outro.

alandefreitas commented:
Se a coisa funciona, vc manda um PR com o que funciona.

Se a coisa não funciona, vc 1) faz um PR com tudo que funciona até então, 2) **para tudo** e 3) fazemos um PR draft para resolver **este** problema até resolvermos. Não faz sentido acumular erro em cima de código que já não funciona.

thiagomacg committed:
Unit tests : corrigindo comparação constant::e(), abs(), cos(), sin(), cosh(), sinh() com racionais, booleanos e inteiros

thiagomacg committed:
Unit tests : erro product::collect()
child_nodes da erro de referência após a absorção do produto

Alteração na absorb_sum_of_sums em summation.cpp

thiagomacg commented on source/sympp/node/operation/product.cpp:
```diff
@@ -271,9 +271,9 @@ namespace sympp {
                 child_nodes_.insert(child_nodes_.end(), p->child_nodes_.begin(),
                                     p->child_nodes_.end());
-                i = child_nodes_.erase(i);
             } else {
```
Boa noite Alan, 

Poderia me ajudar neste trecho do collect() ? 
Este erro esta relacionado ao teste: 
REQUIRE(logA.evaluate_sym(bool_val, int_val, doub_val) == logA); 

Verifica a avaliação simbólica de logA(90,180). 

São realizadas as conversões,  e por fim a simplificação chega a função de collect(). 
Child_Nodes contém dois elementos  (1*1)  e 4,49 que passa a possuir três elementos 1 e 1 e 4,49.

A instrução i = child_nodes_.erase(i); é executada corretamente.

Mas em seguida ao acessar o processo item do iterador é lançado erro. 
Notei que outros pontos também o erase() é utilizado.

O que poderia ser ?

alandefreitas replied:
Bem, este trecho faz conversões do tipo `a * (a * a) * a -> a * a * a * a`. Ele passa por cada elemento filho do product e, se o elemento filho também for product ele copia seus elementos nós para o elemento pai.

O problema é que vc alterou a linha `i = child_nodes_.erase(i);` de lugar. Ou seja, vc apaga o filho antes de copiar os elementos dele para o pai.

`s`, `p` e `i` estão todos apontando para o mesmo element. Só muda o tipo do ponteiro. Ele deveria copiar os elementos de `s` (ou `p` ou `i`) para o pai. Na linha 274 vc deleta o elemento. Na linha 275, vc está tentando copiar os filhos de um elemento que acabou de deletar.

thiagomacg commented on source/sympp/node/operation/summation.cpp:
```diff
@@ -395,9 +396,24 @@ namespace sympp {
+                        ++i;
+                    }
+                }
+            }
```
Boa noite, 

Considerando o teste : REQUIRE(sum.evaluate_sym(bool_val, int_val, doub_val).simplify() == 13.3);
Neste ponto fiz a alteração, já que a simplificação anterior retornava zero

No modo anterior o absorb estava retornando child_nodes vazio. 
A soma a simplificar é : 1 + 1 + 18.3 - 7 

A lista de internal_sums gerada era : < 1, 1, 18.3, -7 >

A lista de child_nodes neste ponto inicia vazia, assim não há o ponteiro para o fim da lista child_nodes.end() e o insert não add os elementos: 

  child_nodes_.insert(child_nodes_.end(),                              std::make_move_iterator(internal_sum.begin()),  std::make_move_iterator(internal_sum.end()));

Para o caso de teste ficou resolvido, realizando o absorb. 

Porém na sequência da simplificação acontece novamento o erro apresentado no collect() do product.cpp, falha após o erase no iterador.

thiagomacg committed:
Unit tests : erro product::collect()
alteração em collect() para coletar os itens dentro de um produto e em seguida não gerar produto de produto.
Ainda sim apresenta erro, pois entra em loop ao tentar simplificar um number que é um product. Linha 385

alandefreitas reviewed (commented):
Ótimo. Tente remover os casos que não funcionam para que o código passe nos unit tests e vá os adicionando aos poucos na medida que vão passando a funcionar. Para conseguirmos fazer o merge mesmo.

alandefreitas commented on source/sympp/node/operation/product.cpp:
```diff
@@ -265,20 +265,35 @@ namespace sympp {
-                i = child_nodes_.erase(i);
-                child_nodes_.insert(child_nodes_.end(), p->child_nodes_.begin(),
-                                    p->child_nodes_.end());
+                // child_nodes_.insert(child_nodes_.end(), p->child_nodes_.begin(),
```
Remover comentários pessoais

alandefreitas commented on tests/unit_tests/sym_functions.cpp:
```diff
@@ -215,15 +216,16 @@ TEST_CASE("Log") {
+    // REQUIRE(logA.evaluate(bool_val, int_val, doub_val) ==   logB.evaluate(bool_val, int_val, doub_val));
 
+    // Com erro
     REQUIRE(logA.evaluate_sym(bool_val, int_val, doub_val) == logA);
```
Tente enviar só o código que passar nos unit tests e vá corrigindo na medida que cria novos testes. O que for comentário pessoal, favor remover antes de enviar ao repositório. O que for comentário relevante, como um bug que descobriu mas não vai corrigir agora, aí sim pode escrever que é um erro e deixar a linha que causaria o erro comentada.

alandefreitas commented on tests/unit_tests/sym_functions.cpp:
```diff
@@ -279,13 +282,23 @@ TEST_CASE("Product") {
     sym absC = sym(sympp::abs(C));
 
-    sym prod = A * absC;
+    //sym prod = A * absC;
```
Remover comentários pessoais.
